### PR TITLE
Feat: Introduced scikit-learn based SGD Classifier model for Incremental Changes

### PR DESCRIPTION
**Pull request by Kaushl2208**
### Why SGD?

As with the Idea introduction to Data-Pipeline, Safaa needs an incremental learning approach

**Support Vector Machines (SVM)**

Pros:

- High Performance: SVMs are known for their effectiveness in high-dimensional spaces and can be very accurate with the right kernel.

- Well-Suited for Non-Linear Problems: Using kernel tricks, SVMs can handle non-linear relationships.

Cons:
- Training Time: SVMs can be very slow to train, especially with large datasets.

- Scalability: SVMs don't scale well with the number of samples and can become impractical with very large datasets.

- Static Model: Once an SVM is trained, incorporating new data typically requires retraining from scratch, which is not ideal for incremental learning.

**Stochastic Gradient Descent (SGD)**

Pros:
- Incremental Learning: SGD is inherently well-suited for incremental learning, allowing you to update the model with new data without retraining from scratch.

- Scalability: Can handle large datasets efficiently since it processes one (or a few) samples at a time.

- Speed: Generally faster to train compared to SVM, especially with large datasets.

Cons:
- Hyperparameter Sensitivity: Requires careful tuning of hyperparameters like the learning rate and regularization strength.

- Convergence Issues: Might not always converge to the global optimum, particularly with a poor choice of learning rate.

### What does the comparison result look like?
Both the models were trained using same pre-processing techniques and same data (_false_positive_detection_dataset.csv_)

SVM:
image

SGD:
image

Accuracy seems almost similar. There are not a lot of difference in the functionality and it comes with _partial_fit()_ approach that can provide incremental learning.

**Integration & Testing**

Tested one manual copyright: 
image

cc: @Hero2323

### CODE REVIEW

File: Safaa/src/safaa/Safaa.py [hunk 1]
```diff
@@ -241,12 +256,13 @@ def train_false_positive_detector_model(self, data, labels):
         # Preprocess the data before training
         preprocessed_data = self.preprocess_data(data)
         # Fit the vectorizer to the preprocessed data
-        vectorized_data = self.vectorizer.fit_transform(preprocessed_data)
+        vectorized_data = self.vectorizer.transform(preprocessed_data)
```
Comment (Hero2323) [inline Safaa/src/safaa/Safaa.py:259]: Changing from 
`vectorized_data = self.vectorizer.fit_transform(preprocessed_data)`
to
`vectorized_data = self.vectorizer.transform(preprocessed_data)`
makes sense now to align the code with incremental learning. The main problem is vectorizers don't support this feature, so the pipeline needs to take this point into consideration, unless we don't plan on updating the corpus often.

Maybe adding a separate function for training the vectorizer from scratch might be of benefit, where the pipeline can accumulate some data first then it can be trained on that entire combined corpus every once in a while.

Not necessarily a request to change anything. Just something to keep in mind.

Comment (Kaushl2208) [reply Safaa/src/safaa/Safaa.py:259]: Yes I also found that issue yesterday, when I was shifting. I think we need to create a stage for vectorizer training as well. Which we might need to think about :)
How we can keep the track of the data :'( is the issue we need to resolve.

**Review by GMishx — CHANGES_REQUESTED** [review]
A minor documentation change is required for function `train_false_positive_detector_model()`. It currently says it trains the model from **scratch** which will no longer be the case once this PR is merged.

**Review by GMishx — APPROVED** [review]
Changes looks good. Tested, working as expected.
